runtime/lua/vim/filetype.lua

```diff
@@ -1226,6 +1227,7 @@ local pattern = {
   [".*/%.config/systemd/user/.*%.d/.*%.conf"] = "systemd",
   [".*/etc/systemd/system/.*%.d/.*%.conf"] = "systemd",
   [".*%.t%.html"] = "tilde",
+  [".?tmux.*%.conf.*"] = "tmux",
```

Comment:
The leading `.` needs to be escaped. And the pattern in `filetype.vim` doesn't have a `*` at the end.

```suggestion
  ["%.?tmux.*%.conf"] = "tmux",
```

Reply:
Thank you for the review.

I doubled checked `filetype.vim` and I found that there are two entries for tmux.

- [line 1887](https://github.com/neovim/neovim/blob/master/runtime/filetype.vim#L1887)
  ```vim
  au BufNewFile,BufRead {.,}tmux*.conf		setf tmux
  ```

- [line 2367](https://github.com/neovim/neovim/blob/master/runtime/filetype.vim#L2367)
  ```vim
  au BufNewFile,BufRead {.,}tmux*.conf*		setf tmux
  ```

In line 2367, the pattern does have `*` at the end. I wonder which one is correct.

Reply:
Good catch, it looks like this is indeed intentional: https://github.com/vim/vim/commit/e519eb41c1c12836b2d12aeb703bb04c7618a724

Reply:
<https://github.com/vim/vim/pull/8971#issuecomment-939315001>

> Adding a star at the end of a filetype pattern is tricky, because it
> matches other things, such as ".gz" or ".old".  It should always be
>checked last.

Reply:
This can be complicated. Can `filetype.lua` currently achieve this?

Reply:
I see. If that is the case should I make two entries such as the following

```lua
["%.?tmux.*%.conf"] = "tmux",
["%.?tmux.*%.conf.*"] = "tmux",
```

where the second entry is put further down to avoid cases such as `.tmux.conf.gz` is recognized as `tmux` filetype instead of `gzip` filetype.

Or is one entry `["%.?tmux.*%.conf.*"] = "tmux"` enough?

Reply:
I think Lua tables are unordered?

Reply:
I just started learning Lua recently so I didn't know about Lua tables being unordered.

To keep in line `filetype.vim`, I think one entry of `["%.?tmux.*%.conf.*"] = "tmux"` is more appropriate than `["%.?tmux.*%.conf"] = "tmux"`.

Is my understanding correct?

Reply:
I'm not sure about this either.

Reply:
Hmm, I wonder why Bram didn't want to use `s:StarSetf` then? (e have our own `starsetf` function to match those). From `filetype.vim`:

```
" Extra checks for when no filetype has been detected now.  Mostly used for
" patterns that end in "*".  E.g., "zsh*" matches "zsh.vim", but that's a Vim
" script file.
" Most of these should call s:StarSetf() to avoid names ending in .gz and the
" like are used.
```

In any case, we can enforce priority in the patterns with the priority key. So what we want is

```lua
["%.?tmux.*%.conf"] = "tmux",
["%.?tmux.*%.conf.*"] = { "tmux", { priority = -1 } }
```

Reply:
Thank you for the clarification. I will push the new changes.

Reply:
Lua tables are indeed unordered; one of the goals of this refactor is to move away from relying on (hard to follow) ordering and instead using explicit priorities, as in
```lua
[".*/etc/foo/.*%.conf"] = { "dosini", { priority = 10 } },
```

Reply:
I see. Thank you for the explanation.